### PR TITLE
Implement Euler Ancestral Discrete scheduler

### PR DESCRIPTION
**Pull request by mspronesti**
Hi @LaurentMazare,
this PR aims at integrating the [Euler Ancestral Discrete Scheduler](https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_euler_ancestral_discrete.py) into this repository, solving the second task mentioned in #23 .

It includes all the features implemented in HF's Python version.

All the considerations made in my last PR, regarding the need to generalize the examples, still apply here.

### CODE REVIEW

**Comment by sssemil** [comment]
image


Tried this with sd1.5 and this prompt:
```
    "n_steps": 21,
    "seed": 1573789502,
    "prompt": "A rusty robot holding a fire torch.",
    "guidance_scale": 7.5
```
I think something is wrong here, or in my code.

**Comment by mspronesti** [comment]
Hi @sssemil,
I tried with HF python diffusers, using your same config (expect the seed) and here are some images with different seeds. To me they seem aligned with the one you obtained with my code.

image

image

image

**Comment by sssemil** [comment]
Hm, but at the same time, this is what I get with python snippet you wrote in another thread:
image

**Comment by mspronesti** [comment]
Did you change the scheduler and the number of inference steps (some of those snippets have 1000)? I tried several times with that same snippet and the appropriate scheduler. Here are other examples I just produced

image

image

image


I leave the exact snippet hereby, to make sure we're using the same one

```py
import torch
from diffusers import StableDiffusionPipeline, EulerAncestralDiscreteScheduler

pipe = StableDiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16)

pipe.scheduler = EulerAncestralDiscreteScheduler()

pipe = pipe.to("cuda")
prompt = "A very rusty robot holding a fire torch."
image = pipe(prompt, num_inference_steps=21).images[0]  
image
```

**Comment by sssemil** [comment]
I think I ran it with 50 steps before, here's with 21:
image

Welp, then it's similar results.

**Comment by mspronesti** [comment]
Happy to help :+1:

**Review by LaurentMazare — APPROVED** [review]

**Comment by LaurentMazare** [comment]
Merged, thanks for all the scheduler hard work!

**Comment by mspronesti** [comment]
Thanks for merging (maybe forgot to push After git merge)?

**Comment by LaurentMazare** [comment]
Oh right sorry about that!
